### PR TITLE
added cache for did-resolver

### PR DESCRIPTION
**Pull request by Cxxshyy**
Wanted to use this new branch i created for the Pull request if thats ok

### CODE REVIEW

File: src/did/did-resolver.ts [hunk 1]
```diff
@@ -46,7 +51,13 @@ export class DidResolver {
       throw new Error(`${didMethod} DID method not supported`);
     }
 
-    const resolutionResult = await didResolver.resolve(did);
+    const resolutionResult = await this.cache.get(did) !== undefined ? await this.cache.get(did): await didResolver.resolve(did);
+    const isCached = await this.cache.get(did) !== undefined;
```
Comment (thehenrytsai) [inline src/did/did-resolver.ts:55]: `await this.cache.get(did)` is called 3 times, please reduce it to just once

File: src/did/did-resolver.ts [hunk 2]
```diff
@@ -46,7 +51,13 @@ export class DidResolver {
...
+    if (!isCached){
+      await this.cache.set(did,resolutionResult);
+    } else {
+
+    }
```
Comment (thehenrytsai) [inline src/did/did-resolver.ts:60]: No need to have an empty `else` block.

File: tests/did/did-resolver.spec.ts [hunk 3]
```diff
@@ -20,6 +21,20 @@ describe('DidResolver', () => {
     });
     await didResolver.resolve(did);
 
+    expect(ionDidResolveSpy.called).to.be.true;
+  });
+  it('should pick the right DID resolver based on DID method name and account for passed in cache', async () => {
```
Comment (thehenrytsai) [inline tests/did/did-resolver.spec.ts:26]: The test seems unrelated to the addition of a cache and does not cover the code added. The test should be along the lines of:

"should read from cache and not from resolver if cache contains the DID resolution result"

**Review by thehenrytsai — CHANGES_REQUESTED** [review]
Thanks @Cxxshyy! Some suggestions.

**Comment by Cxxshyy** [comment]
not to sure whether you meant for the test to be like this but is this what you meant?

**Comment by Cxxshyy** [comment]
im confused why integrity check failed as i just tested and they all passed

File: src/did/did-resolver.ts [hunk 4]
```diff
@@ -50,13 +50,10 @@ export class DidResolver {
     if (!didResolver) {
       throw new Error(`${didMethod} DID method not supported`);
     }
-
-    const resolutionResult = await this.cache.get(did) !== undefined ? await this.cache.get(did): await didResolver.resolve(did);
-    const isCached = await this.cache.get(did) !== undefined;
-    if (!isCached){
+    const isCached = await this.cache.get(did);
```
Comment (thehenrytsai) [inline src/did/did-resolver.ts:53]:
```suggestion
    const cachedResolutionResult = await this.cache.get(did);
```

File: src/did/did-resolver.ts [hunk 5]
```diff
@@ -50,13 +50,10 @@ export class DidResolver {
...
+    const resolutionResult = isCached !== undefined ? isCached : await didResolver.resolve(did);
```
Comment (thehenrytsai) [inline src/did/did-resolver.ts:54]:
```suggestion
    const resolutionResult = cachedResolutionResult ?? await didResolver.resolve(did);
```

**Comment by thehenrytsai** [comment]
> not to sure whether you meant for the test to be like this but is this what you meant?

Not quite, see PR https://github.com/TBD54566975/dwn-sdk-js/commit/4d5d983c23cb5f72459a98c3a84128b5af6913c6 for test example, basically you want to make sure resolution for the SAME DID doesn't call the resolver twice (i.e. once hitting the resolver, and once hitting the cache)

> im confused why integrity check failed as i just tested and they all passed

Most likely because more code has been merged into `main` branch, and there is conflict with your PR (looks like just README.md for now). In case you didn't know, you can either run git commands and fix conflicts locally, or you can use the github UI to resole the conflicts.

**Comment by michaelneale** [comment]
this in reasonable shape now to merge?

**Comment by Cxxshyy** [comment]
@michaelneale need some amendments first ill do them now

**Comment by Cxxshyy** [comment]
@thehenrytsai sorry if i still dont understand but what i done was check cache to see if result is there and if not call resolver then store that result in cache

**Comment by thehenrytsai** [comment]
@Cxxshyy, I have gone ahead and made the changes I was thinking off to your PR: https://github.com/Cxxshyy/dwn-sdk-js/pull/1, please take a look and approve, then I can merge!

**Comment by thehenrytsai** [comment]
> @thehenrytsai it is often possible to push directly to the @Cxxshyy#1 branch (yes - the remote branch! usually github allows it with forked PRs, which I know sounds weird but works the same as a branch based PR).

Wow, this is super weird (but also super cool) instead, would've have thought one would need explicit permission from the repo owner, I guess it's implied!! Merging it!

**Pull request merged**